File: src/activecontext/plugins/transport.py
"""CAP Transport — stdio implementation (StdioTransport).

JSON-RPC 2.0 over stdio pipes. This is the default CAP transport
implementation; see cap_transport.py for the abstract protocol.

Low-level bidirectional transport for communication with plugin servers.
Each message is a single JSON object followed by a newline.

The transport is bidirectional:
- Host → Server: requests (with id) and notifications (no id)
- Server → Host: responses (matching id) and push notifications

Threading model:
- One asyncio reader task reads from the server's stdout
- Incoming messages are dispatched: responses to pending futures,
  notifications to the registered callback
- Writes go directly to the server's stdin (serialized under a lock)
"""

from __future__ import annotations

import asyncio
import json
import logging
import os
from typing import Any, Callable

from activecontext.plugins.wire import (
    ErrorCodes,
    to_jsonrpc_notification,
    to_jsonrpc_request,
)

logger = logging.getLogger(__name__)

# Type for the notification callback: (method, params) -> None
NotificationCallback = Callable[[str, dict[str, Any]], None]


class PluginTransportError(Exception):
    """Error in the plugin transport layer."""


class JsonRpcError(PluginTransportError):
    """JSON-RPC error response from the server."""

    def __init__(self, code: int, message: str, data: Any = None) -> None:
        self.code = code
        self.data = data
        super().__init__(f"JSON-RPC error {code}: {message}")


class PluginTransport:
    """JSON-RPC 2.0 transport over stdio pipes.

    Manages a subprocess with stdin/stdout pipes for bidirectional
    JSON-RPC communication. Supports both request/response (with id)
    and fire-and-forget notifications.

    Usage:
        transport = PluginTransport(command=["python", "-m", "my_plugin"])
        await transport.start()
        result = await transport.send_request("initialize", {...})
        await transport.send_notification("shutdown", {})
        await transport.stop()
    """

    def __init__(
        self,
        command: list[str],
        env: dict[str, str] | None = None,
        cwd: str | None = None,
        on_notification: NotificationCallback | None = None,
    ) -> None:
        """Initialize transport configuration.

        Args:
            command: Command and arguments to spawn the plugin server.
            env: Additional environment variables (merged with os.environ).
            cwd: Working directory for the subprocess.
            on_notification: Callback for server → host notifications.
        """
        self._command = command
        self._env = env
        self._cwd = cwd
        self._on_notification = on_notification

        self._process: asyncio.subprocess.Process | None = None
        self._reader_task: asyncio.Task[None] | None = None
        self._next_id = 1
        self._pending: dict[int | str, asyncio.Future[Any]] = {}
        self._write_lock = asyncio.Lock()
        self._started = False
        self._stopping = False

    @property
    def is_running(self) -> bool:
        """Whether the transport subprocess is running."""
        return (
            self._started
            and not self._stopping
            and self._process is not None
            and self._process.returncode is None
        )

    async def start(self) -> None:
        """Spawn the plugin server subprocess and start the reader.

        Raises:
            PluginTransportError: If the subprocess fails to start.
        """
        if self._started:
            raise PluginTransportError("Transport already started")

        # Merge environment
        merged_env = dict(os.environ)
        if self._env:
            merged_env.update(_expand_env_vars(self._env))

        try:
            self._process = await asyncio.create_subprocess_exec(
                *self._command,
                stdin=asyncio.subprocess.PIPE,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
                env=merged_env,
                cwd=self._cwd,
            )
        except (OSError, FileNotFoundError) as e:
            raise PluginTransportError(
                f"Failed to start plugin server: {self._command[0]}: {e}"
            ) from e

        self._started = True
        self._reader_task = asyncio.create_task(
            self._read_loop(), name="cap-transport-reader"
        )
        logger.info(
            "CAP transport started: %s (pid=%s)",
            " ".join(self._command),
            self._process.pid,
        )

    async def stop(self) -> None:
        """Stop the transport and terminate the subprocess.

        Cancels pending requests with an error. Waits for the
        subprocess to exit gracefully, then kills if needed.
        """
        if not self._started or self._stopping:
            return

        self._stopping = True

        # Cancel all pending requests
        for future in self._pending.values():
            if not future.done():
                future.set_exception(
                    PluginTransportError("Transport shutting down")
                )
        self._pending.clear()

        # Cancel reader task
        if self._reader_task and not self._reader_task.done():
            self._reader_task.cancel()
            try:
                await self._reader_task
            except asyncio.CancelledError:
                pass

        # Terminate subprocess
        if self._process and self._process.returncode is None:
            try:
                self._process.terminate()
                try:
                    await asyncio.wait_for(self._process.wait(), timeout=5.0)
                except asyncio.TimeoutError:
                    self._process.kill()
                    await self._process.wait()
            except ProcessLookupError:
                pass  # Already dead

        self._started = False
        self._stopping = False
        logger.info("CAP transport stopped")

    async def send_request(
        self, method: str, params: Any = None, timeout: float = 30.0
    ) -> Any:
        """Send a JSON-RPC request and wait for the response.

        Args:
            method: RPC method name.
            params: Parameters (dataclass or dict).
            timeout: Seconds to wait for response.

        Returns:
            The result field from the JSON-RPC response.

        Raises:
            JsonRpcError: If the server returns an error response.
            PluginTransportError: If the transport is not running.
            asyncio.TimeoutError: If no response within timeout.
        """
        if not self.is_running:
            raise PluginTransportError("Transport not running")

        request_id = self._next_id
        self._next_id += 1

        msg = to_jsonrpc_request(method, params, id=request_id)
        future: asyncio.Future[Any] = asyncio.get_event_loop().create_future()
        self._pending[request_id] = future

        try:
            await self._write(msg)
            return await asyncio.wait_for(future, timeout=timeout)
        except asyncio.TimeoutError:
            self._pending.pop(request_id, None)
            raise
        except Exception:
            self._pending.pop(request_id, None)
            raise

    async def send_notification(self, method: str, params: Any = None) -> None:
        """Send a JSON-RPC notification (no response expected).

        Args:
            method: RPC method name.
            params: Parameters (dataclass or dict).

        Raises:
            PluginTransportError: If the transport is not running.
        """
        if not self.is_running:
            raise PluginTransportError("Transport not running")

        msg = to_jsonrpc_notification(method, params)
        await self._write(msg)

    async def _write(self, msg: dict[str, Any]) -> None:
        """Write a JSON-RPC message to stdin."""
        assert self._process is not None
        assert self._process.stdin is not None

        data = json.dumps(msg, separators=(",", ":")) + "\n"
        async with self._write_lock:
            self._process.stdin.write(data.encode("utf-8"))
            await self._process.stdin.drain()

    async def _read_loop(self) -> None:
        """Background task: read JSON-RPC messages from stdout."""
        assert self._process is not None
        assert self._process.stdout is not None

        try:
            while not self._stopping:
                line = await self._process.stdout.readline()
                if not line:
                    # EOF — subprocess exited
                    break

                line_str = line.decode("utf-8").strip()
                if not line_str:
                    continue

                try:
                    msg = json.loads(line_str)
                except json.JSONDecodeError as e:
                    logger.warning("CAP: invalid JSON from server: %s", e)
                    continue

                self._dispatch(msg)

        except asyncio.CancelledError:
            pass
        except Exception as e:
            logger.error("CAP reader error: %s", e)
        finally:
            # EOF or error: fail all pending requests
            if not self._stopping:
                for future in self._pending.values():
                    if not future.done():
                        future.set_exception(
                            PluginTransportError("Server connection lost")
                        )
                self._pending.clear()

    def _dispatch(self, msg: dict[str, Any]) -> None:
        """Route an incoming message to the correct handler."""
        if "id" in msg and "method" not in msg:
            # Response (has id, no method)
            self._handle_response(msg)
        elif "method" in msg and "id" not in msg:
            # Notification (has method, no id)
            self._handle_notification(msg)
        elif "method" in msg and "id" in msg:
            # Request from server (has both method and id)
            # This is a host API call — route to notification callback
            # which the connection layer handles as a reverse request
            self._handle_server_request(msg)
        else:
            logger.warning("CAP: unrecognized message: %s", msg)

    def _handle_response(self, msg: dict[str, Any]) -> None:
        """Handle a JSON-RPC response."""
        msg_id = msg["id"]
        future = self._pending.pop(msg_id, None)
        if future is None:
            logger.warning("CAP: response for unknown id: %s", msg_id)
            return

        if "error" in msg:
            err = msg["error"]
            future.set_exception(
                JsonRpcError(
                    code=err.get("code", ErrorCodes.INTERNAL_ERROR),
                    message=err.get("message", "Unknown error"),
                    data=err.get("data"),
                )
            )
        else:
            future.set_result(msg.get("result"))

    def _handle_notification(self, msg: dict[str, Any]) -> None:
        """Handle a server → host notification."""
        method = msg["method"]
        params = msg.get("params", {})
        if self._on_notification:
            try:
                self._on_notification(method, params)
            except Exception as e:
                logger.error("CAP notification handler error: %s", e)
        else:
            logger.debug("CAP: unhandled notification: %s", method)

    def _handle_server_request(self, msg: dict[str, Any]) -> None:
        """Handle a server → host request (host API call).

        These are bidirectional requests where the server calls the host.
        Routed through the notification callback with the full message
        so the connection layer can send a response.
        """
        method = msg["method"]
        params = msg.get("params", {})
        request_id = msg["id"]

        if self._on_notification:
            # Pass the full message including id so the connection
            # layer can respond
            try:
                self._on_notification(
                    method,
                    {"_request_id": request_id, **params},
                )
            except Exception as e:
                logger.error("CAP host API handler error: %s", e)
        else:
            logger.warning("CAP: unhandled host API request: %s", method)

    async def send_response(self, result: Any, request_id: int | str) -> None:
        """Send a JSON-RPC response to a server request.

        Used by the connection layer to respond to host API calls.

        Args:
            result: Result value (dataclass or dict).
            request_id: The id from the server's request.
        """
        from activecontext.plugins.wire import to_jsonrpc_response

        msg = to_jsonrpc_response(result, id=request_id)
        await self._write(msg)

    async def send_error_response(
        self,
        code: int,
        message: str,
        request_id: int | str | None,
        data: Any = None,
    ) -> None:
        """Send a JSON-RPC error response to a server request.

        Args:
            code: Error code.
            message: Human-readable error message.
            request_id: The id from the server's request.
            data: Optional additional error data.
        """
        from activecontext.plugins.wire import to_jsonrpc_error

        msg = to_jsonrpc_error(code, message, id=request_id, data=data)
        await self._write(msg)


# Backward-compatible alias: StdioTransport is the concrete name,
# PluginTransport remains importable for existing code.
StdioTransport = PluginTransport


def _expand_env_vars(env: dict[str, str]) -> dict[str, str]:
    """Expand ${VAR} references in environment variable values."""
    result = {}
    for key, value in env.items():
        if isinstance(value, str) and "${" in value:
            # Expand all ${VAR} references
            import re

            def replacer(match: re.Match[str]) -> str:
                return os.environ.get(match.group(1), "")

            result[key] = re.sub(r"\$\{([^}]+)\}", replacer, value)
        else:
            result[key] = value
    return result

File: sdks/typescript/src/transport.ts
/**
 * Stdio transport for CAP JSON-RPC communication.
 *
 * Reads newline-delimited JSON from stdin, writes to stdout.
 * Implements the reference stdio binding from the CAP transport spec.
 *
 * @module transport
 */

import * as readline from "readline";
import type {
  JsonRpcRequest,
  JsonRpcNotification,
  JsonRpcResponse,
  JsonRpcErrorResponse,
  JsonRpcMessage,
} from "./types";

// ---------------------------------------------------------------------------
// Message classification
// ---------------------------------------------------------------------------

/** Classify an incoming JSON-RPC message by its shape. */
export type IncomingMessage =
  | { kind: "request"; message: JsonRpcRequest }
  | { kind: "notification"; message: JsonRpcNotification }
  | { kind: "response"; message: JsonRpcResponse }
  | { kind: "error"; message: JsonRpcErrorResponse };

function classifyMessage(raw: Record<string, unknown>): IncomingMessage | null {
  const hasId = "id" in raw && raw.id !== undefined;
  const hasMethod = "method" in raw && typeof raw.method === "string";
  const hasResult = "result" in raw;
  const hasError = "error" in raw;

  if (hasMethod && hasId) {
    return { kind: "request", message: raw as unknown as JsonRpcRequest };
  }
  if (hasMethod && !hasId) {
    return { kind: "notification", message: raw as unknown as JsonRpcNotification };
  }
  if (hasResult && hasId) {
    return { kind: "response", message: raw as unknown as JsonRpcResponse };
  }
  if (hasError && hasId) {
    return { kind: "error", message: raw as unknown as JsonRpcErrorResponse };
  }

  return null;
}

// ---------------------------------------------------------------------------
// Pending request tracking
// ---------------------------------------------------------------------------

interface PendingRequest {
  resolve: (result: unknown) => void;
  reject: (error: Error) => void;
  timer?: ReturnType<typeof setTimeout>;
}

// ---------------------------------------------------------------------------
// StdioTransport
// ---------------------------------------------------------------------------

/** Handler for incoming JSON-RPC requests from the host. */
export type RequestHandler = (
  method: string,
  params: unknown,
  id: number | string
) => Promise<unknown>;

/** Handler for incoming JSON-RPC notifications from the host. */
export type NotificationHandler = (method: string, params: unknown) => void;

/**
 * StdioTransport reads JSON-RPC messages from stdin and writes to stdout.
 *
 * Supports:
 * - Sending requests with response tracking (via pending futures).
 * - Sending notifications (no response expected).
 * - Sending responses to incoming requests.
 * - Dispatching incoming messages to registered handlers.
 *
 * Framing: newline-delimited JSON (NDJSON). Each message is a single
 * compact JSON object followed by `\n`.
 */
export class StdioTransport {
  private rl: readline.Interface | null = null;
  private nextId = 1;
  private pending = new Map<number | string, PendingRequest>();
  private running = false;
  private requestHandler: RequestHandler | null = null;
  private notificationHandler: NotificationHandler | null = null;
  private readonly defaultTimeout: number;

  /**
   * @param defaultTimeout - Default timeout in milliseconds for requests (default: 30000).
   */
  constructor(defaultTimeout = 30000) {
    this.defaultTimeout = defaultTimeout;
  }

  /** Whether the transport is actively reading from stdin. */
  get isRunning(): boolean {
    return this.running;
  }

  /** Register a handler for incoming requests from the host. */
  onRequest(handler: RequestHandler): void {
    this.requestHandler = handler;
  }

  /** Register a handler for incoming notifications from the host. */
  onNotification(handler: NotificationHandler): void {
    this.notificationHandler = handler;
  }

  /**
   * Start reading from stdin.
   *
   * This begins the read loop. Messages are dispatched to registered
   * handlers as they arrive.
   */
  start(): void {
    if (this.running) return;
    this.running = true;

    this.rl = readline.createInterface({
      input: process.stdin,
      crlfDelay: Infinity,
    });

    this.rl.on("line", (line: string) => {
      this.handleLine(line);
    });

    this.rl.on("close", () => {
      this.handleClose();
    });
  }

  /**
   * Stop the transport and release resources.
   *
   * Fails all pending requests with a transport error. Idempotent.
   */
  stop(): void {
    if (!this.running) return;
    this.running = false;

    if (this.rl) {
      this.rl.close();
      this.rl = null;
    }

    // Fail all pending requests
    for (const [id, pending] of this.pending) {
      if (pending.timer) clearTimeout(pending.timer);
      pending.reject(new Error(`Transport stopped, request ${id} abandoned`));
    }
    this.pending.clear();
  }

  /**
   * Send a JSON-RPC request and wait for the response.
   *
   * @param method - The RPC method name.
   * @param params - Parameters object.
   * @param timeout - Timeout in milliseconds (default: transport default).
   * @returns The result from the response.
   * @throws Error if the request times out or the transport is closed.
   */
  sendRequest(
    method: string,
    params?: unknown,
    timeout?: number
  ): Promise<unknown> {
    return new Promise((resolve, reject) => {
      if (!this.running) {
        reject(new Error("Transport is not running"));
        return;
      }

      const id = this.nextId++;
      const timeoutMs = timeout ?? this.defaultTimeout;

      const timer = setTimeout(() => {
        this.pending.delete(id);
        reject(new Error(`Request ${method} (id=${id}) timed out after ${timeoutMs}ms`));
      }, timeoutMs);

      this.pending.set(id, { resolve, reject, timer });

      const message: JsonRpcRequest = {
        jsonrpc: "2.0",
        method,
        id,
        ...(params !== undefined ? { params } : {}),
      };

      this.writeLine(message);
    });
  }

  /**
   * Send a JSON-RPC notification (no response expected).
   *
   * @param method - The RPC method name.
   * @param params - Parameters object.
   */
  sendNotification(method: string, params?: unknown): void {
    if (!this.running) return;

    const message: JsonRpcNotification = {
      jsonrpc: "2.0",
      method,
      ...(params !== undefined ? { params } : {}),
    };

    this.writeLine(message);
  }

  /**
   * Send a JSON-RPC success response.
   *
   * @param result - The result value.
   * @param id - The request ID this responds to.
   */
  sendResponse(result: unknown, id: number | string): void {
    if (!this.running) return;

    const message: JsonRpcResponse = {
      jsonrpc: "2.0",
      id,
      result,
    };

    this.writeLine(message);
  }

  /**
   * Send a JSON-RPC error response.
   *
   * @param code - Error code (from ErrorCodes).
   * @param message - Human-readable error description.
   * @param id - The request ID this responds to (null for parse errors).
   * @param data - Optional additional error data.
   */
  sendErrorResponse(
    code: number,
    message: string,
    id: number | string | null,
    data?: unknown
  ): void {
    if (!this.running) return;

    const errorResp: JsonRpcErrorResponse = {
      jsonrpc: "2.0",
      id: id as number | string,
      error: {
        code,
        message,
        ...(data !== undefined ? { data } : {}),
      },
    };

    this.writeLine(errorResp);
  }

  // -----------------------------------------------------------------------
  // Internal
  // -----------------------------------------------------------------------

  private writeLine(message: JsonRpcMessage): void {
    const json = JSON.stringify(message);
    process.stdout.write(json + "\n");
  }

  private handleLine(line: string): void {
    const trimmed = line.trim();
    if (!trimmed) return; // Ignore empty lines per spec

    let raw: Record<string, unknown>;
    try {
      raw = JSON.parse(trimmed) as Record<string, unknown>;
    } catch {
      // Malformed JSON -- log to stderr, do not crash
      process.stderr.write(`[CAP] Invalid JSON on stdin: ${trimmed.substring(0, 100)}\n`);
      return;
    }

    const classified = classifyMessage(raw);
    if (!classified) {
      process.stderr.write(`[CAP] Unclassifiable message: ${trimmed.substring(0, 100)}\n`);
      return;
    }

    switch (classified.kind) {
      case "request":
        this.handleRequest(classified.message);
        break;
      case "notification":
        this.handleNotification(classified.message);
        break;
      case "response":
        this.handleResponse(classified.message);
        break;
      case "error":
        this.handleErrorResponse(classified.message);
        break;
    }
  }

  private handleRequest(msg: JsonRpcRequest): void {
    if (!this.requestHandler) {
      this.sendErrorResponse(
        -32601,
        `No handler for method: ${msg.method}`,
        msg.id
      );
      return;
    }

    this.requestHandler(msg.method, msg.params, msg.id)
      .then((result) => {
        this.sendResponse(result, msg.id);
      })
      .catch((err: Error) => {
        this.sendErrorResponse(
          -32603,
          err.message || "Internal error",
          msg.id
        );
      });
  }

  private handleNotification(msg: JsonRpcNotification): void {
    if (this.notificationHandler) {
      this.notificationHandler(msg.method, msg.params);
    }
  }

  private handleResponse(msg: JsonRpcResponse): void {
    const pending = this.pending.get(msg.id);
    if (!pending) return;

    this.pending.delete(msg.id);
    if (pending.timer) clearTimeout(pending.timer);
    pending.resolve(msg.result);
  }

  private handleErrorResponse(msg: JsonRpcErrorResponse): void {
    const pending = this.pending.get(msg.id);
    if (!pending) return;

    this.pending.delete(msg.id);
    if (pending.timer) clearTimeout(pending.timer);

    const err = new Error(
      `JSON-RPC error ${msg.error.code}: ${msg.error.message}`
    );
    (err as unknown as Record<string, unknown>).code = msg.error.code;
    (err as unknown as Record<string, unknown>).data = msg.error.data;
    pending.reject(err);
  }

  private handleClose(): void {
    // stdin closed (EOF) -- treat as shutdown
    this.stop();
  }
}
